[ci][testing] Add automated periodic Trivy tests

gmaiorov opened the pull request:
## Description
<!---
  Describe your changes in detail.

  Please let users know if your feature influences critical cluster components
  (restarts of ingress-controllers, control-plane, Prometheus, etc).
-->
This PR adds CVE test CI workflow that uses [Trivy](https://github.com/aquasecurity/trivy).
Base images used in build process, built Deckhouse image & module images are subject to test.

[GitHub CI workflow template](https://github.com/deckhouse/deckhouse/blob/8b502923ec32db74d6e811c3a0d65afdadbb4db2/.github/workflow_templates/cve-daily.yml) contained in this PR allows to configure whether this workflow should:
- work on test repositories using `$enableWorkflowOnTestRepos` flag;
- test all Deckhouse release channels or just latest images based on `main` branch using `$testAllReleaseChannels`;
- allow to test Deckhouse images independently of base images, or depend on it's successful tests using `$dependOnBaseImageTest`.

Workflow was tested using [deckhouse-test-2 repository](https://github.com/deckhouse/deckhouse-test-2/) with release channel branches and corresponding tags populated.
Example available in [following run](https://github.com/deckhouse/deckhouse-test-2/actions/runs/3337076596/).

Workflow is executed periodically using crontab.

## Why do we need it, and what problem does it solve?
<!---
  This is the most important paragraph.
  You must describe the main goal of your feature.

  If it fixes an issue, place a link to the issue here.

  If it fixes an obvious bug, please tell users about the impact and effect of the problem.
-->
Needed in #2679 (the other part is to update base images).
## What is the expected result?
<!---
  How can one check these changes after applying?  

  Describe, what (resource, state, event, etc.) MUST or MUST NOT change/happen after applying these changes.
-->
Periodic Trivy tests are added.

## Checklist
- [ ] The code is covered by unit tests.
- [x] e2e tests passed.
- [ ] Documentation updated according to the changes.
- [ ] Changes were tested in the Kubernetes cluster manually.

## Changelog entries
<!---
  Describe the changes so they will be included in a release changelog.

  Find examples and documentation below, or visit the instruction page on the repo wiki
  https://github.com/deckhouse/deckhouse/wiki/How-to-add-to-changelog
-->
```changes
section: ci
type: feature
summary: Add automated periodic Trivy tests
impact_level: low
```

<!---
`impact_level: default` adds to changelog as usual, this is the default that can be omitted
`impact_level: high`    something important for users, the impact will be copied to "Know Before Update" section
`impact_level: low`     omitted in changelog YAML; note there is `type:chore` for chores

Tip for the section field:

  - <kebab-case of a module>, e.g. "cloud-provider-aws", "node-manager"
  - "ci", has forced low impact
  - "docs", includes website changes, should have low impact
  - "candi"
  - "deckhouse-controller"
  - "dhctl"
  - "global-hooks"
  - "go_lib"
  - "helm_lib"
  - "jq_lib"
  - "shell_lib"
  - "testing", has forced low impact
  - "tools", has forced low impact

Find changed sections:

gh pr diff   $PULL_REQUEST_NUMBER   |
  egrep "^([+]{3} b|[-]{3} a)/" |
  cut -d/ -f2- |
  sed 's#^ee/##' |
  sed 's#^fe/##' |
  sed 's#^modules/##' |
  sed 's#[0-9][0-9][0-9]-##' |
  egrep -v 'Makefile' |       # add file exclusion here
  cut -d/ -f1 |
  sort |
  uniq

Find all possible sections (excluding ci):

node -e 'console.log(require("./.github/scripts/js/changelog-find-sections.js")().join("\n"))'
-->

deckhouse-BOaTswain commented:
Aye, aye, @gmaiorov. I've started the workflow for label 'e2e/run/gcp'...
<!-- workflow_start -->

RomanenkoDenys commented on .github/workflow_templates/cve-daily.yml:
```diff
@@ -0,0 +1,85 @@
+{!{/* Jobs for each release channel */}!}
+{!{- $channel_names := slice "main" -}!}
+{!{- if $testAllReleaseChannels -}!}
+{!{-   $channel_names = $channel_names | coll.Append (slice "alpha" "beta" "early-access" "stable" "rock-solid") | coll.Flatten -}!}
```
we should check releases (like 1.37) instead of release channels.

gmaiorov replied:
I've implemented per-release testing using Github Actions matrix strategy and git describe to map deployed tags to release channel branches.

RomanenkoDenys commented on tools/cve/release.sh:
```diff
@@ -47,7 +49,8 @@ function __main__() {
   tags=$(docker run --rm "$REPO:$TAG" cat /deckhouse/modules/images_tags.json)
 
-  trivy image --timeout 10m --severity=$SEVERITY "$REPO:$TAG"
+  trivy imag123e --timeout 10m --severity=$SEVERITY --exit-code 1 "$REPO:$TAG" || ret=$?
```
```suggestion
  trivy image --timeout 10m --severity=$SEVERITY --exit-code 1 "$REPO:$TAG" || ret=$?
```

RomanenkoDenys reviewed (dismissed):
small changes

diafour commented on .github/workflow_templates/cve-daily.yml:
```diff
@@ -0,0 +1,85 @@
+{!{-   if eq $ref_name "main" -}!}
+{!{-     $repo = "dev-registry.deckhouse.io/sys/deckhouse-oss" -}!}
+{!{-   else -}!}
+{!{-     $repo = "registry.deckhouse.io/deckhouse/fe" -}!}
```
"else" branch seems redundant because of $repo initilization.

gmaiorov replied:
This code was removed from template during migration to matrix strategy instead of fixed jobs.

diafour commented on .github/workflow_templates/cve-daily.yml:
```diff
@@ -0,0 +1,85 @@
+{!{ tmpl.Exec "checkout_full_step"           $ctx | strings.Indent 2 }!}
+{!{ tmpl.Exec "login_dev_registry_step"      $ctx | strings.Indent 2 }!}
+{!{ tmpl.Exec "login_readonly_registry_step" $ctx | strings.Indent 2 }!}
+{!{ tmpl.Exec "link_bin_step"                     | strings.Indent 2 }!}
```
link_bin_step and unlink_bin_step can be moved here as-is (without "define").

diafour replied:
Let's leave it as is in favor of #2789

gmaiorov commented:
Rebased onto main due to conflicts in Makefile.

deckhouse-BOaTswain commented:
Aye, aye, @gmaiorov. I've started the workflow for label 'e2e/run/gcp'...
<!-- workflow_start -->

gmaiorov commented on Makefile:
```diff
@@ -48,8 +48,10 @@ else ifeq ($(PLATFORM_NAME), arm64)
 endif
 
 # Set host arch & OS for golang-based programs, e.g. Prometheus
+ifneq (, $(shell which go))
```
Removed in 5fd10276b5fd84b8c769578d1ee7d8d0464a6440.

gmaiorov commented on Makefile:
```diff
@@ -39,7 +40,6 @@ else ifeq ($(PLATFORM_NAME), arm64)
 	TRDL_ARCH = arm64
 endif
 
-
```
Added in d70592147cbb7d66fbc7c69a670ada3c704b2a6d.

diafour commented on tools/cve/tag_map.sh:
```diff
@@ -0,0 +1,40 @@
+
+set -Eeo pipefail
+shopt -s failglob
+
```
There is no description what is the purpose of this script. What data it takes and what it returns.

gmaiorov replied:
I've added a description for tag_map.sh in bf15eeb417a4481e70bd9e47406737751e8d2f61.

gmaiorov replied:
I've added a description for tag_map.sh in bf15eeb417a4481e70bd9e47406737751e8d2f61.

diafour commented on .github/workflows/cve-daily.yml:
```diff
@@ -0,0 +1,373 @@
+  - cron: '0 23 * * 1-5'
+  workflow_dispatch:
+
+# Always run a single job at a time.
+# Note: Concurrency is currently in beta and subject to change.
+# https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#concurrency
```
It may be deleted. `concurrency` is not a beta now.

diafour commented on .github/workflows/cve-daily.yml:
```diff
@@ -0,0 +1,373 @@
+        run: |
+          REPO="registry.deckhouse.io/deckhouse/fe"
+          echo "REPO=${REPO}" >> $GITHUB_ENV
+          TAG=$(echo "${{ matrix.tag }}" | grep -Eo "^\S+")
```
Using gha expressions (`${{ ... }}`) inside the shell script text is not reliable. It may corrupt the script text if expression result contains quotes, newlines or other fancy symbols. It is better pass them using env:

```
- name: ...
  env:
    matrix_tag: ${{ matrix.tag }}
  run: |
    echo "${matrix_tag}" | ...
```

For example, see [passing job output and a context value via envs](https://github.com/deckhouse/deckhouse/blob/main/.github/workflows/deploy-alpha.yml#L357-L362). Also see passing gha expressions results as [json-encoded strings](https://github.com/deckhouse/deckhouse/blob/main/.github/workflows/deploy-alpha.yml#L767-L770).

gmaiorov replied:
I've applied your suggestion about GHA expressions in c2832026961731ceb2aa38eaadbefa25da5d18b3.

gmaiorov replied:
I've applied your suggestion about GHA expressions in c2832026961731ceb2aa38eaadbefa25da5d18b3.

diafour commented on .github/workflows/cve-daily.yml:
```diff
@@ -0,0 +1,373 @@
+          echo "REPO=${REPO}" >> $GITHUB_ENV
+          TAG=$(echo "${{ matrix.tag }}" | grep -Eo "^\S+")
+          echo "TAG=${TAG}" >> $GITHUB_ENV
+      - name: Run CVE tests
+        run: |
+          echo "⚓️ 📤 Checking out candi/image_versions.yml from ${TAG}..."
+          git checkout --force "${TAG}" -- candi/image_versions.yml
+          echo "⚓️ 🏎 Running CVE tests on ${TAG}..."
+          make cve-report
```
Is it correct that this job should not run if previous job is skipped? In this case previous job ('Set repo and tag for matrix run') can return output and this job ('Run CVE tests') may have `if` that checks this output. For example, see steps "Check registry"/"Login to registry".

gmaiorov replied:
Nope, we shouldn't skip this job simultaneously with a previous one.
Required env vars could be set manually in workflow definition (e.g. for main branch) or via `Set repo and tag for matrix run` step (when executed with matrix strategy).

diafour reviewed (changes requested):
I've left some comments

gmaiorov commented on .github/workflows/cve-daily.yml:
```diff
@@ -0,0 +1,373 @@
+          ln -s ~/deckhouse-bin-cache bin
+      # </template: link_bin_step>
+      - name: Set repo and tag for matrix run
+        if: matrix.tag != ''
```
Nope, it's not relevant, but it's used to distinct whether job is running using matrix strategy.
It comes from single `cve_tests` template that creates both `test_cve_report_main` and `test_cve_report_channels`.

It is possible to completely remove this step for main branch during template rendering, but it seems that `cve_tests` could look a bit clearer if I rely on GHA condition. Which approach should I take?

gmaiorov replied:
I've reworked workflow in edc156cbc4fe121d8b5d8e13d5f4b43cc07f336a and removed GHA conditions in favor of templating.

gmaiorov commented:
I've reworked workflow in edc156cbc4fe121d8b5d8e13d5f4b43cc07f336a and removed GHA conditions in favor of templating.
Example of updated workflow run available at [second test repo](https://github.com/deckhouse/deckhouse-test-2/actions/runs/3330395973/) (same as in updated description).

deckhouse-BOaTswain commented:
Aye, aye, @gmaiorov. I've started the workflow for label 'e2e/run/gcp'...
<!-- workflow_start -->

deckhouse-BOaTswain commented:
Aye, aye, @gmaiorov. I've started the workflow for label 'e2e/run/gcp'...
<!-- workflow_start -->
